### PR TITLE
ct: add an option to start with an initial snapshot of the input

### PR DESCRIPTION
**Pull request by danhhz**
This adds a `WITH` option to the `CONTINUAL TASK` syntax (including the `TRANSFORM USING` sugar) to run the CT statements on an initial snapshot of the input contents. It is meant to be the default, but there are some issues to shake out with how it interacts with self-referential CTs (filed as https://github.com/MaterializeInc/database-issues/issues/8699).

The motivation is twofold:
- It will probably be common that the user wants this initial backfill. For example, our system builtin CTs do.
- The initial as_of of a CT is not guaranteed to be strict serializable (same for MVs). With the initial snapshot, this is not easily observable, but without it, it's easy to end up in counter-intuitive situations like: insert, create ct, and then see the insert in the ct even though it happened "before". It's possible that some users will want the no snapshot behavior, so leave it in behind an option.

Touches https://github.com/MaterializeInc/database-issues/issues/8427

### Motivation

  * This PR adds a feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by danhhz** [comment]
Okay @teskje @antiguru, this is the PR I mentioned that I'd love to get merged for this week's release. The demo prep took longer than I expected, so I didn't get this polished before the end of my day. It definitely needs a PR/commit message, and there's a bug I'm still shaking out in self-referential CTs (a hang), but it's far enough along that you could get an initial review pass in.

File: src/adapter/src/coord/sequencer/inner/create_continual_task.rs
Comment (teskje) [inline src/adapter/src/coord/sequencer/inner/create_continual_task.rs:113]: This is not related to/required by the rest of the PR, right?

File: src/compute/src/render/continual_task.rs
Comment (teskje) [inline src/compute/src/render/continual_task.rs:192]: Nit: Merge this into the last branch?

**Review by teskje — APPROVED** [review]
LGTM, provided you can make CI happy. An adapter review would be good, for the `mz-sql` changes.

**Review by antiguru — APPROVED** [review]
+1 to Jan's comments, otherwise LGTM.

**Review by ParkMyCar — APPROVED** [review]
Adapter bits LGTM!

File: src/adapter/src/coord/sequencer/inner/create_continual_task.rs
Comment (danhhz) [reply src/adapter/src/coord/sequencer/inner/create_continual_task.rs:113]: Yup. Added a comment here

File: src/compute/src/render/continual_task.rs
Comment (danhhz) [reply src/compute/src/render/continual_task.rs:192]: Done!

File: src/sql/src/plan/statement/ddl.rs
Comment (danhhz) [reply src/sql/src/plan/statement/ddl.rs:2778]: Done!

**Comment by danhhz** [comment]
The hang I was hitting with self-referential CTs ended up being pretty thorny and I don't have a good resolution yet. Typed it up as part of https://github.com/MaterializeInc/database-issues/issues/8699. Given that it doesn't put us in any worse of a place, gonna flip the default on the snapshot option back to false (exactly the behavior we had before this PR) until we get that resolved and merge this.

TFTRs!

**Pull request merged**
